File: Loenn/entities/wingedAdder.lua
local entities = require("entities")

local winged_adder = {}

winged_adder.name = "WingedHelper/WingedAdder"
winged_adder.depth = -100000

local outlineColor = { 0.737, 0.941, 0.960, 0.75 }
local fillColor = { 0.537, 0.698, 0.756, 0.45 }


local getAllSids = function ()
    return {}
end

if entities and entities.registeredEntities then
    getAllSids = function ()
        local ret = {}
        for k,v in pairs(entities.registeredEntities) do
            table.insert(ret, k)
        end
        table.sort(ret)
        return ret
    end
end


winged_adder.fieldInformation = function(entity) return {
    selectedTypes = {
        fieldType = "list",
        elementSeparator = ",",
        elementDefault = "",
        elementOptions = {
            options = getAllSids(),
            searchable = true,
        },
    },
    direction = {
        options = { "Up", "Down", "Left", "Right" },
        editable = false,
    },
    leftWingTint = {
        fieldType = "color",
        elementDefault = "FFFFFF",
    },
    rightWingTint = {
        fieldType = "color",
        elementDefault = "FFFFFF",
    },
} end

winged_adder.fieldOrder = function(entity) return{ 
    "x", "y", "width", "height", "leftWingXOffset", "rightWingXOffset", "leftWingYOffset", "rightWingYOffset","moveDelay", 
    "flySpeed", "direction", "selectedTypes", "leftWingTint", "rightWingTint", "blacklist", "inAreaRange", "actorsOnly", "collidablesOnly", "isHeavyWings", "allowInteractions", "disableCollisions", "rainbowWings",
}end



winged_adder.placements ={
    name = "WingedAdder",
    data = {
        width = 16,
        height = 16,
        selectedTypes = "",
        direction = "Up",
        leftWingTint = "FFFFFF",
        rightWingTint = "FFFFFF",
        blacklist = false,
        inAreaRange = true,
        actorsOnly = false,
        collidablesOnly = true,
        isHeavyWings = false,
        allowInteractions = true,
        disableCollisions = false,
        rainbowWings = false,
        moveDelay = 1.0,
        flySpeed = 25.0,
        leftWingXOffset = 0,
        leftWingYOffset = 0,
        rightWingXOffset = 0,
        rightWingYOffset = 0,
    },
}


winged_adder.fillColor = fillColor
winged_adder.borderColor = outlineColor


return winged_adder

File: Source/WingedAdder.cs
using System;
using System.Collections.Generic;
using Celeste.Mod.Entities;
using Microsoft.Xna.Framework;
using Monocle;

namespace Celeste.Mod.WingedHelper;

[CustomEntity("WingedHelper/WingedAdder")]
public class WingedAdder : Entity
{
    List<Entity> entities = new();
    EntityFilter filter;
    
    bool mustBeInArea;
    bool actorsOnly;
    bool collidablesOnly;

    bool disableCollisions;
    bool allowInteractions;
    
    bool heavyWings;
    float moveDelay;
    int upSpeed;
    FlyDirection direction;
    Vector2 leftWingOffset = new();
    Vector2 rightWingOffset = new();
    
    string leftColor;
    string rightColor;
    bool rainbow;
    bool flySounds;
    bool flapSounds;
    
    int dashsToActivate;
    string counterColor;
    Vector2 counterOffset = new();
    
    
    public WingedAdder(EntityData data, Vector2 offset) : base(data.Position + offset)
    {
        Collider = new Hitbox(data.Width, data.Height);
        filter = EntityFilter.CreateFromData(data);
        mustBeInArea = data.Bool("inAreaRange");
        heavyWings = data.Bool("isHeavyWings");
        moveDelay = data.Float("moveDelay");
        upSpeed = data.Int("flySpeed");
        actorsOnly = data.Bool("actorsOnly");
        collidablesOnly = data.Bool("collidablesOnly");
        direction = data.Enum("direction", FlyDirection.Up);
        leftWingOffset.X = data.Int("leftWingXOffset");
        leftWingOffset.Y = data.Int("leftWingYOffset");
        rightWingOffset.X = data.Int("rightWingXOffset");
        rightWingOffset.Y = data.Int("rightWingYOffset");
        disableCollisions = data.Bool("disableCollisions");
        allowInteractions = data.Bool("allowInteractions");
        leftColor = data.Attr("leftWingTint", "FFFFFF");
        rightColor = data.Attr("rightWingTint", "FFFFFF");
        rainbow = data.Bool("rainbowWings");
        flySounds = data.Bool("flySounds");
        flapSounds = data.Bool("flapSounds");
        dashsToActivate = data.Int("dashsToActivate");
        counterColor = data.Attr("counterTintColor", "FFFFFF");
        counterOffset.X = data.Int("counterXOffset");
        counterOffset.Y = data.Int("counterYOffset");
    }

    public override void Awake(Scene scene)
    {
        base.Awake(scene);
        
        foreach (Entity entity in scene.Entities)
        {
            if (entity == this || entity.Get<WingComponent>() != null) continue;
            if ((!mustBeInArea || Collider.Collide(entity.Position)) && filter.Matches(entity))
            {
                entities.Add(entity);
            }
        }
        
        foreach (Entity entity in entities)
        {
            if (entity.Get<WingComponent>() != null) continue;
            bool isActor = entity is Actor;
            bool hasCollider = entity.Collider != null;
            if ((!actorsOnly || (isActor && actorsOnly)) && (!collidablesOnly || (hasCollider && collidablesOnly)))
            {
                Logger.Log(LogLevel.Info, "WingedHelper", $"Adding WingComponent to {entity}");
                WingComponent wingComp = new WingComponent(moveDelay, upSpeed, heavyWings, direction, isActor, hasCollider, leftWingOffset, rightWingOffset, disableCollisions, 
                    allowInteractions, leftColor, rightColor, rainbow, flySounds, flapSounds, dashsToActivate, counterColor, counterOffset);
                entity.Add(wingComp);
            }
        }
        
        RemoveSelf();
    }
}